lib/stdlib/src/erl_tar.erl

```diff
@@ -225,809 +279,1581 @@ month(10) -> "Oct";
+    BaseHeader = #tar_header{name=Name,
+                             mtime=Fi#file_info.mtime,
+                             atime=Fi#file_info.atime,
+                             ctime=Fi#file_info.ctime,
+                             mode=Fi#file_info.mode,
+                             uid=Fi#file_info.uid,
+                             gid=Fi#file_info.gid,
+                             typeflag=?TYPE_REGULAR},
+    do_fileinfo_to_header(BaseHeader, Fi, Link).
+
+do_fileinfo_to_header(Header, #file_info{size=Size,type=regular}, _Link) ->
+    Header#tar_header{size=Size,typeflag=?TYPE_REGULAR};
+do_fileinfo_to_header(#tar_header{name=Name}=Header,
+                      #file_info{type=directory}, _Link) ->
+    Header#tar_header{name=Name++"/",typeflag=?TYPE_DIR};
+do_fileinfo_to_header(Header, #file_info{type=symlink}, Link) ->
+    Header#tar_header{typeflag=?TYPE_SYMLINK,linkname=Link};
+do_fileinfo_to_header(Header, #file_info{type=device,mode=Mode}=Fi, _Link)
+  when (Mode band ?S_IFMT) =:= ?S_IFCHR ->
+    Header#tar_header{typeflag=?TYPE_CHAR,
+                      devmajor=Fi#file_info.major_device,
+                      devminor=Fi#file_info.minor_device};
+do_fileinfo_to_header(Header, #file_info{type=device,mode=Mode}=Fi, _Link)
+  when (Mode band ?S_IFMT) =:= ?S_IFBLK ->
+    Header#tar_header{typeflag=?TYPE_BLOCK,
+                      devmajor=Fi#file_info.major_device,
+                      devminor=Fi#file_info.minor_device};
+do_fileinfo_to_header(Header, #file_info{type=other,mode=Mode}, _Link)
+  when (Mode band ?S_IFMT) =:= ?S_FIFO ->
+    Header#tar_header{typeflag=?TYPE_FIFO};
+do_fileinfo_to_header(Header, Fi, _Link) ->
+    {error, {invalid_file_type, Header#tar_header.name, Fi}}.
+
+is_ascii(Str) when is_list(Str) ->
+    not lists:any(fun (Char) -> Char >= 16#80 end, Str);
+is_ascii(<<>>) ->
+    true;
+is_ascii(<<C,_Rest/binary>>) when C >= 16#80 ->
+    false;
+is_ascii(<<_, Rest/binary>>) ->
+    is_ascii(Rest).
+
+to_ascii(Str) when is_list(Str) ->
+    case is_ascii(Str) of
+        true ->
+            Str;
+        false ->
+            lists:filter(fun (Char) -> Char < 16#80 end, Str)
+    end;
+to_ascii(Bin) when is_binary(Bin) ->
+    to_ascii(Bin, <<>>).
+to_ascii(<<>>, Acc) ->
+    Acc;
+to_ascii(<<C, Rest/binary>>, Acc) when C < 16#80 ->
+    to_ascii(Rest, <<Acc/binary,C>>);
+to_ascii(<<_, Rest/binary>>, Acc) ->
+    to_ascii(Rest, Acc).
+
+is_header_only_type(?TYPE_SYMLINK) -> true;
+is_header_only_type(?TYPE_LINK)    -> true;
+is_header_only_type(?TYPE_DIR)     -> true;
+is_header_only_type(_) -> false.
+
+posix_to_erlang_time(Sec) ->
+    OneMillion = 1000000,
+    Time = calendar:now_to_datetime({Sec div OneMillion, Sec rem OneMillion, 0}),
+    erlang:universaltime_to_localtime(Time).
+
+foldl_read(TarName, Fun, Accu, #read_opts{}=Opts)
+  when is_function(Fun,4) ->
+    case catch open(TarName, [read|Opts#read_opts.open_mode]) of
```

Comment:
Use `try`...`of` here instead `catch`.

Error exceptions should be passed through, not converted to `{error,Reason}`. That is, only throws should be catched.